### PR TITLE
add cache

### PR DESCRIPTION
**Pull request by kittowned**
Added cache mechanism to address the issues https://github.com/yoavbls/pretty-ts-errors/issues/58 and https://github.com/yoavbls/pretty-ts-errors/issues/36

After adding the cache and benchmarking with performance.now() I was able to get roughly 10x performance improvement on large TS files.
It's not a perfect solution however, since the cache uses a combination of the diagnostic range and ts error code to differentiate between diagnostics, when modifying a TS file by pushing code down a line all diagnostics below that line change their ranges and as a result they are not found in the cache. Any ideas to improve this are very welcome.

I've also taken several profiles of the extension and most of the time is spent in markdown generation.
Specifically in dedent and prettify libraries. When replacing those libraries with simple functions that do nothing and just return the strings extension performance improves significantly.

I'm not sure how to proceed from here, any advice will be greatly appreciated 😃 .

### CODE REVIEW

**Comment by yoavbls** [comment]
Hey @kittowned, Thank you for creating this PR.
Solving the performance issues is super important, and you profiled it which is excellent. Thank you very much for that!
I want to suggest another method for the caching instead of the Cache class.
I thought about the caching keys a lot. As you find out, using the code + line ranges could be problematic in some cases because the lines are changing a lot. most of the errors would be with the same 4-5 error codes.

I think the best option would be a simple JavaScript Map with the whole error message string as keys. The keys would be very long, but it should be performant to find them and it doesn't require any calculations to generate them or find the values.

e.g:
```typescript
const cache = new Map();
let formattedMessage = cache.get(diagnostic.message)

if(!formattedMessage) {
  // All the expensive formatting is happening here in formatDiagnosticMessage
  formattedMessage = formatDiagnosticMessage(diagnostic.message);
  cache.set(diagnostic.message, formattedMessage)
}
```
and if we want to limit the memory footprint and set the max size for the Map  
we can remove stale values by adding something like this after the `cache.set`

```typescript
if(cache.size > 20) { // or any size we want
  const firstCacheKey = cache.keys().next().value
  cache.delete(firstCacheKey)
}
```
Let me know what you think about it 🙂

**Comment by kittowned** [comment]
Hi @yoavbls !
Thank you very much for going over my PR, I really appreciate it!

I really like your elegant solution, I've went ahead and implemented it on my machine.
In terms of performance this is definitely the way to go, we get all the previous gains from the cache and changing code lines doesn't invalidate the cache.

However, I did find a case that could be problematic, when there are more than one unique diagnostic messages, they show up on the first element that matches the message, please see screenshot with my contrived example:
image

I'm not sure this is a deal breaker because I don't see many users encountering this, but on the other hand it will crop up in places and could cause a bit of confusion. Something to consider.
Please let me know your thoughts :)

**Comment by yoavbls** [comment]
@kittowned Thank you! 
I'm not sure precisely what is causing this behavior. I think we can use the cached formatted message separately for every error and avoid these duplications.|
Can you commit your changes, please? I'll take a look 🙂

**Comment by kittowned** [comment]
Hi @yoavbls .
Sure thing, committed.

File: src/extension.ts [hunk 1]
```diff
@@ -36,23 +38,31 @@ export function activate(context: ExtensionContext) {
...
+                formatDiagnostic(converter.asDiagnostic(diagnostic), prettify)
+              );
+              
+              markdownString.isTrusted = true;
+              markdownString.supportHtml = true;
+              
+              formattedMessage = {
+                range: diagnostic.range,
+                contents: [markdownString],
+              };
```
Comment (yoavbls) [inline src/extension.ts:58]: We need to save only the formatted message in the cache without the range:

```typescript
formattedMessage = markdownString;
cache.set(diagnostic.message, formattedMessage);
```
and then in push we'll add the range:
```typscript
items.push({
 range: diagnostic.range,
 contents: [formattedMessage],
});
```

That way, the message will not appear twice or in the wrong places 🙂

**Review by yoavbls — COMMENTED** [review]
Thanks @kittowned!
Read my last comment, it'll solve the problem you mentioned 🙂

**Comment by kittowned** [comment]
Hi @yoavbls !
Thank you for the detailed explanation.
You are definitely right, I didn't consider separating the formatted message from the range in this way 😅 .
I've pushed a new commit with these changes.
Do you think we need to add a cache limit at this time?

**Review by yoavbls — APPROVED** [review]
@kittowned looks great!
You can add a limit of 100~ if you will :)
Or I'll merge it and we'll add the limit later. Whatever you want 🙂

**Comment by kittowned** [comment]
Hey @yoavbls , I've added a limit of 100 as you advised.
Please merge at your leisure 😄

**Comment by yoavbls** [comment]
Thank you!! 🙂

**Comment by WesleyYue** [comment]
@yoavbls Would love to have this fix. Could you cut a new release with this fix? If not I can build myself as well.

**Comment by yoavbls** [comment]
@WesleyYue  I plan to release a new one at the end of the following week but I can make a one-off version with this fix, do you want that?

**Comment by WesleyYue** [comment]
@yoavbls yeah that would be awesome. I've been turning this extension on only when I need it due to the performance issues. Really appreciate your work on this!

**Comment by yoavbls** [comment]
@WesleyYue v0.5 is out 🙂
https://github.com/yoavbls/pretty-ts-errors/releases/tag/v0.5.0

**Pull request merged**